### PR TITLE
Added sents property to Span for Spans spanning over several sentences

### PR DESCRIPTION
**Pull request by nrodnova**
Added `sents` property to `Span` for `Spans` spanning over several sentences

## Description

Added `sents` property to `Span` for `Spans` spanning over several sentences. It works similarly to `Doc.sents` returning a generator of sentences. I also added tests and updated documentation.

### Types of change

enhancement, tests, documentation

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy/tests/doc/test_span.py [hunk 1]
```diff
@@ -34,7 +34,6 @@ def doc(en_tokenizer):
         lemmas=lemmas,
     )
 
-
```
Comment (svlandeg) [inline spacy/tests/doc/test_span.py:37]: We format all files in the main repo with `black` - you can find more information [here](https://github.com/explosion/spaCy/blob/master/CONTRIBUTING.md#code-conventions). If you run `black` you'll see that this newline will be added in again ;-)

File: spacy/tests/doc/test_span.py [hunk 2]
```diff
@@ -98,6 +98,11 @@ def test_spans_span_sent(doc, doc_not_parsed):
     assert doc[:2].sent.root.text == "is"
     assert doc[:2].sent.text == "This is a sentence."
     assert doc[6:7].sent.root.left_edge.text == "This"
+    assert doc[0: len(doc)].sent == list(doc.sents)[0]
```
Comment (svlandeg) [inline spacy/tests/doc/test_span.py:101]: I'd add an additional line to clarify the difference between `span.sent` and `span.sents`:
```suggestion
    assert doc[0 : len(doc)].sent == list(doc.sents)[0]
    assert list(doc[0 : len(doc)].sents) == list(doc.sents)
```

File: spacy/tests/doc/test_span.py [hunk 3]
```diff
@@ -434,3 +439,59 @@ def test_span_with_vectors(doc):
     # single-token span with vector
     assert_array_equal(ops.to_numpy(doc[10:11].vector), [-1, -1, -1])
     doc.vocab.vectors = prev_vectors
+
+def test_span_sents(doc, doc_not_parsed) :
+    # Entire doc
+    span = Span(doc, 0, len(doc))
+    sentences = list(span.sents)
+    assert len(sentences) == 3
```
Comment (svlandeg) [inline spacy/tests/doc/test_span.py:447]: To shorten the code, I'd remove the var from all of these checks:
```suggestion
    assert len(list(span.sents)) == 3
```

File: spacy/tests/doc/test_span.py [hunk 4]
```diff
@@ -434,3 +439,59 @@ def test_span_with_vectors(doc):
     # single-token span with vector
     assert_array_equal(ops.to_numpy(doc[10:11].vector), [-1, -1, -1])
     doc.vocab.vectors = prev_vectors
+
+def test_span_sents(doc, doc_not_parsed) :
```
Comment (svlandeg) [inline spacy/tests/doc/test_span.py:443]: Could you rewrite this check using `@pytest.mark.parametrize`? Basically you'd define the offsets (we can hardcode `len(doc)`) and hardcode the expected number of sentences, and you'd only have to write the `assert` logic once.

The remaining part of this test can then be run as a separate function.

File: spacy/tests/doc/test_span.py [hunk 5]
```diff
@@ -98,6 +98,11 @@ def test_spans_span_sent(doc, doc_not_parsed):
     assert doc[:2].sent.root.text == "is"
     assert doc[:2].sent.text == "This is a sentence."
     assert doc[6:7].sent.root.left_edge.text == "This"
+    assert doc[0: len(doc)].sent == list(doc.sents)[0]
+
+    with pytest.raises(ValueError) :
+        doc_not_parsed[:2].sent
```
Comment (svlandeg) [inline spacy/tests/doc/test_span.py:104]: I think we can remove this part from this test, as it's already included in a later test (where it's using `span = Span(doc_not_parsed, 0, 3)`)

File: spacy/tokens/span.pyx [hunk 6]
```diff
@@ -422,6 +422,40 @@ cdef class Span:
         else:
             raise ValueError(Errors.E030)
 
+    @property
+    def sents(self):
+        """Obtain the sentences that contains this span. If the given span
+        crosses sentence boundaries, returns all sentences it is a part of
```
Comment (svlandeg) [inline spacy/tokens/span.pyx:428]:
```suggestion
        """Obtain the sentences that contain this span. If the given span
        crosses sentence boundaries, return all sentences it is a part of.
```

File: spacy/tokens/span.pyx [hunk 7]
```diff
@@ -422,6 +422,40 @@ cdef class Span:
...
+
+        RETURNS (Iterable[Span]): The genetator of sentences the span is a part of.
```
Comment (svlandeg) [inline spacy/tokens/span.pyx:430]:
```suggestion
        RETURNS (Iterable[Span]): All sentences that the span is a part of.
```

File: website/docs/api/span.md [hunk 8]
```diff
@@ -506,17 +506,26 @@ sent = span.sent
...
-> span = doc[1:3]
-> assert span.sent.text == "Give it back!"
+> span = doc[2:4]
+> assert len(span.sents) == 2
 > ```
 
 | Name        | Description                                             |
 | ----------- | ------------------------------------------------------- |
-| **RETURNS** | The sentence span that this span is a part of. ~~Span~~ |
+| **RETURNS** | A generator yielding sentences this `Span` is a part of ~~Iterable[Span]~~ |
```
Comment (svlandeg) [inline website/docs/api/span.md:528]: FYI - for formatting the documentation, we use `prettier`.

File: website/docs/api/span.md [hunk 9]
```diff
@@ -506,17 +506,26 @@ sent = span.sent
 sent = doc[sent.start : max(sent.end, span.end)]
 ```
 
+## Span.sents {#sents tag="property" model="sentences"}
```
Comment (svlandeg) [inline website/docs/api/span.md:509]: I think something went wrong while copy/pasting. The Example and table for `span.sent` has been deleted. Could you restore this?

**Review by svlandeg — CHANGES_REQUESTED** [review]
Thanks a lot for this PR! I agree that this property would make sense to have. I particularly like all the tests you've wrote to ensure the functionality works as intended :-)

I left some reviewing comments, and will mark this as "changes requested" for our internal administration. I think these are mainly small things, and when addressed, we should be able to merge this!

File: spacy/tests/doc/test_span.py (hunk 4 above)
Comment (nrodnova) [reply spacy/tests/doc/test_span.py]: Done

File: spacy/tests/doc/test_span.py (hunk 5 above)
Comment (nrodnova) [reply spacy/tests/doc/test_span.py]: This is a valid test for non-sentecized doc. In the following lines, the sentence boundaries are set manually and the assertions pass without throwing an exception:
```python
   with pytest.raises(ValueError):
        doc_not_parsed[:2].sent

    # test on manual sbd
    doc_not_parsed[0].is_sent_start = True
    doc_not_parsed[5].is_sent_start = True
    assert doc_not_parsed[1:3].sent == doc_not_parsed[0:5]
    assert doc_not_parsed[10:14].sent == doc_not_parsed[5:]

```

File: website/docs/api/span.md (hunk 9 above)
Comment (nrodnova) [reply website/docs/api/span.md:516]: Yes, sorry about it. I corrected.

File: spacy/tests/doc/test_span.py (hunk 5 above)
Comment (svlandeg) [reply spacy/tests/doc/test_span.py]: I'm just trying to be mindful of not repeating the same kind of test too often ;-)

File: spacy/tokens/span.pyx [hunk 10]
```diff
@@ -422,6 +426,48 @@ cdef class Span:
...
+
+        if "sents" in self.doc.user_span_hooks:
+            yield from self.doc.user_span_hooks["sents"](self)
+        elif "sents" in self.doc.user_hooks:
+            for sentence in self.doc.user_hooks["sents"](self.doc) :
+                if sentence.end > self.start :
+                    if sentence.start < self.end or sentence.start == self.start == self.end:
+                        yield sentence
+                    else :
+                        break
```
Comment (svlandeg) [inline spacy/tokens/span.pyx:452]: I believe there is no support for `black` formatting in Cython, so you can't know this, but I'd at least put the `:` directly after the line without space, to match the formatting in other parts of the code base.

File: website/docs/api/span.md [hunk 11]
```diff
@@ -501,11 +501,6 @@ If the span happens to cross sentence boundaries, only the first sentence will
 be returned. If it is required that the sentence always includes the full span,
 the result can be adjusted as such:
 
-```python
-sent = span.sent
-sent = doc[sent.start : max(sent.end, span.end)]
-```
-
```
Comment (svlandeg) [inline website/docs/api/span.md:508]: Why did you remove this part from the `span.sent` documentation?

Comment (nrodnova) [reply website/docs/api/span.md:508]: @svlandeg Devinitely not on purpose :) For some reason, my PyCharm doesn't like editing .md files. Sorry about that, I will make sure I don't screw up the documentation any more :)

File: spacy/tokens/span.pyx (hunk 10 above)
Comment (nrodnova) [reply spacy/tokens/span.pyx]: I updated the formatting

File: spacy/tokens/span.pyx [hunk 12]
```diff
@@ -422,6 +426,48 @@ cdef class Span:
...
+
+        RETURNS (Iterable[Span]): All sentences that the span is a part of.
+
+         DOCS: https://spacy.io/api/span#sents
+        """
+        cdef int start
+        cdef int i
+
+        if not self.doc.has_annotation("SENT_START"):
+            raise ValueError(Errors.E030)
```
Comment (svlandeg) [inline spacy/tokens/span.pyx:442]: This check should come after checking the user hooks - if there is a valid hook we don't have to raise the exception.

Otherwise I think this looks good!

File: website/docs/api/span.md [hunk 13]
```diff
@@ -518,6 +518,27 @@ sent = doc[sent.start : max(sent.end, span.end)]
 | ----------- | ------------------------------------------------------- |
 | **RETURNS** | The sentence span that this span is a part of. ~~Span~~ |
 
+## Span.sents {#sents tag="property" model="sentences"}
```
Comment (svlandeg) [inline website/docs/api/span.md:521]:
```suggestion
## Span.sents {#sents tag="property" model="sentences" new="3.2.1"}
```

**Review by svlandeg — APPROVED** [review]
